quinn/src/delay_queue.rs

```diff
@@ -0,0 +1,442 @@
+            if let Some(timer) = slot.take() {
+                let state = self.timers.remove(timer.0);
+                debug_assert_eq!(state.prev, None, "head of list has no predecessor");
+                if let Some(next) = state.next {
+                    debug_assert_eq!(
+                        self.timers[next.0].prev,
+                        Some(timer),
+                        "successor links to head"
+                    );
+                    self.timers[next.0].prev = None;
+                }
+                *slot = state.next;
+                self.next_tick = state.expiry;
+                return Some(state.value);
+            }
+        }
+        None
+    }
+
+    /// Advance to the start of the first nonempty slot or `now`, whichever is sooner
+    fn advance_towards(&mut self, now: u64) {
+        for level in 0..LEVELS {
+            for slot in range_in_level(level, self.next_tick..=now) {
+                debug_assert!(
+                    now >= slot_start(self.next_tick, level, slot),
+                    "slot overlaps with the past"
+                );
+                if self.levels[level].slots[slot].is_some() {
+                    self.advance_to(level, slot);
+                    return;
+                }
+            }
+        }
+        self.next_tick = now;
+    }
+
+    /// Advance to a specific slot, which must be the first nonempty slot
+    fn advance_to(&mut self, level: usize, slot: usize) {
+        debug_assert!(
+            self.levels[..level]
+                .iter()
+                .all(|level| level.slots.iter().all(|x| x.is_none())),
+            "lower levels are empty"
+        );
+        debug_assert!(
+            self.levels[level].slots[..slot].iter().all(Option::is_none),
+            "lower slots in this level are empty"
+        );
+
+        // Advance into the slot
+        self.next_tick = slot_start(self.next_tick, level, slot);
+
+        // Drain it into the lower levels, if possible
+        if level > 0 {
+            while let Some(timer) = self.levels[level].slots[slot].take() {
+                let next = self.timers[timer.0].next;
+                self.levels[level].slots[slot] = next;
+                if let Some(next) = next {
+                    self.timers[next.0].prev = None;
+                }
+                self.list_unlink(timer);
+                self.schedule(timer);
+            }
+        }
+    }
+
+    /// Link `timer` from the slot associated with its expiry
+    fn schedule(&mut self, timer: Timer) {
+        debug_assert_eq!(
+            self.timers[timer.0].next, None,
+            "timer isn't already scheduled"
+        );
+        debug_assert_eq!(
+            self.timers[timer.0].prev, None,
+            "timer isn't already scheduled"
+        );
+        let (level, slot) = timer_index(self.next_tick, self.timers[timer.0].expiry);
+        // Insert `timer` at the head of the list in the target slot
```

Comment:
Does that mean a freshly added timer will expire before other timers for the same timestamp? I think for fairness reasons it would be good that it expires after those.

Reply:
I'm not sure fairness is the most important concern here. If the system is severely overloaded, LIFO processing allows for some, rather than no, timeouts to be served in a timely manner, which can help reduce cascading failures.

Reply:
That's true. But it can have tricky side effects: E.g. let's say one only prioritizes new connections if the system is highly loaded. Old connections then would time out - but if they are not handled at all the stack won't be able to finalize them and reclaim their resources - which then again means the system might not be able to get out of the high load situation.

Now this problem might not really occur due to how quinn uses the timer: As far as I see in the endpoint code it will handle all expired timers anyway, just in LIFO order.

Reply:
Hmm, good point, we currently drain the whole expired set regardless. We might want to change that in the future for the same reasons we've done so in the send/receive loops, though.

> if they are not handled at all the stack won't be able to finalize them and reclaim their resources

Also an interesting point; I guess I remain ambivalent. One nice thing about LIFO is it doesn't need an extra index, at least.

Reply:
Added a comment to `drive_timers` describing this issue so we'll see it if we make a relevant change, and leaving in LIFO for now for simplicity.